src/main/java/org/broadinstitute/hellbender/tools/exome/DiscoverCopyNumberTriStateSegments.java

```diff
@@ -0,0 +1,484 @@
+            final CopyNumberTriStateSegment s1 = o1.getSegment();
+            final CopyNumberTriStateSegment s2 = o2.getSegment();
+            // Using the index-range.from we make sure we sort first by the contig over
+            // the start position in the same order as contigs are present in
+            // the input data or target collection.
+            final IndexRange ir1 = targets.indexRange(o1.getSegment());
+            final IndexRange ir2 = targets.indexRange(o2.getSegment());
+
+            final int fromCmp = Integer.compare(ir1.from, ir2.from);
+            if (fromCmp != 0) {
+                return fromCmp;
+            }
+            // if fromCmp == 0, they must be in the same contig,
+            // then we can sort based on segment start and then end.
+            final int startCmp = Integer.compare(s1.getStart(), s2.getStart());
+            if (startCmp != 0) {
+                return startCmp;
+            }
+            return Integer.compare(s1.getEnd(), s2.getEnd());
+        };
+
+        return Iterators.mergeSorted(allSegmentRecordIterators, recordComparator);
+    }
+
+    private Map<String, List<CopyNumberTriStateSegment>> calculateBestPathSegments(CopyNumberTriStateHiddenMarkovModel model, ReadCountCollection inputCounts) {
+        final List<String> columnNames = inputCounts.columnNames();
+        final Map<String, List<CopyNumberTriStateSegment>> allSegments = new LinkedHashMap<>(columnNames.size());
+        long recordCount = 0;
+        final List<Target> targets = inputCounts.targets();
+        for (int i = 0; i < columnNames.size(); i++) {
+            final String columnName = columnNames.get(i);
+            final double[] inputValues = inputCounts.counts().getColumn(i);
+            final List<CopyNumberTriStateSegment> bestPathSegmentList = getCopyNumberTriStateSegments(model, targets, inputValues);
+            recordCount += bestPathSegmentList.size();
+            allSegments.put(columnName, bestPathSegmentList);
+        }
+        logger.info(String.format("Found a total of %d segments across %d samples", recordCount, inputCounts.columnNames().size()));
+        return allSegments;
+    }
+
+    /**
+     * Calculates the list of copy-number change segments as the one present in the
+     * most likely segment sequence given a {@link CopyNumberTriStateHiddenMarkovModel model}.
+     * @param model the HMM model to be used to discover copy-number change segments.
+     * @param targets the input targets sequence.
+     * @param inputValues the input coverage values.
+     * @return never {@code null}.
+     */
+    private List<CopyNumberTriStateSegment> getCopyNumberTriStateSegments(final CopyNumberTriStateHiddenMarkovModel model, final List<Target> targets, double[] inputValues) {
+        final double mean = MathUtils.mean(inputValues, 0, inputValues.length);
+        final double invStdev = 1.0 / MathUtils.stddev(inputValues, 0, inputValues.length);
```

Comment:
No need to cache the inverse.  It saves one division per target, but even with 10^6 targets and 10^3 samples we're talking about an optimization that saves less than a second but makes the code harder to read.